### DISCUSSION TITLE
Local function expressions

### DISCUSSION THREAD

**Discussion by portal-chan** [dotnet/csharplang#8734] [Language Ideas]
Don't know if this was already mentioned/is planned/is secretly a terrible idea but I didn't find anything by searching so here goes:

How about making local functions into expressions so that they could be immediately invoked? Kind of how you're able to do in languages with first class functions:

**Before**
```csharp
public static void Main(string[] args)
{
    var result = factorial(5);

    int factorial(int n)
    {
        if (n <= 1) return 1;
        else return n * factorial(n - 1);
    }
}
```

**After**
```csharp
public static void Main(string[] args)
{
    var result = int factorial(int n) 
    {
        if (n <= 1) return 1;
        else return n * factorial(n - 1);
    }(5);
}
```

This would allow one-use recursive local functions to be used at the moment of declaration instead of being left at the top/bottom of a function, arguably resulting in some cleaner code. As far as I'm aware there's no notable drawbacks to it unless there's some implementation ones.

**Comment by jnm2** [comment]
The other possible benefit of an IIFE besides recursion is when the local function is transformed into a state machine (`async` and iterators). This benefit wouldn't require the local function expression to have a name like recursion does.

**Comment by Thaina** [comment]
#249 
#888
